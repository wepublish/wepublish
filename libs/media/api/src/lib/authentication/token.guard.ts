import { AuthGuard } from '@nestjs/passport';

export class TokenAuthGuard extends AuthGuard('token') {}
