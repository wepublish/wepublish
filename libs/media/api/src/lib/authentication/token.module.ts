import {DynamicModule, Module, ModuleMetadata, Provider, Type} from '@nestjs/common'
import {TokenStrategy, TokenConfig, TOKEN_MODULE_OPTIONS} from './token.strategy'

export type TokenOptionsFactory = {
  createTokenOptions(): Promise<TokenConfig> | TokenConfig
}

export interface TokenAsyncOptions extends Pick<ModuleMetadata, 'imports'> {
  useExisting?: Type<TokenOptionsFactory>
  useClass?: Type<TokenOptionsFactory>
  useFactory?: (...args: any[]) => Promise<TokenConfig> | TokenConfig
  inject?: any[]
}

@Module({
  providers: [TokenStrategy],
  exports: [TokenStrategy]
})
export class TokenModule {
  public static register(config: TokenConfig): DynamicModule {
    return {
      module: TokenModule,
      providers: [
        {
          provide: TOKEN_MODULE_OPTIONS,
          useValue: config
        }
      ]
    }
  }

  public static registerAsync(options: TokenAsyncOptions): DynamicModule {
    return {
      module: TokenModule,
      imports: options.imports || [],
      providers: this.createAsyncProviders(options)
    }
  }

  private static createAsyncProviders(options: TokenAsyncOptions): Provider[] {
    if (options.useExisting || options.useFactory) {
      return [this.createAsyncOptionsProvider(options)]
    }

    return [
      this.createAsyncOptionsProvider(options),
      {
        provide: options.useClass!,
        useClass: options.useClass!
      }
    ]
  }

  private static createAsyncOptionsProvider(options: TokenAsyncOptions): Provider {
    if (options.useFactory) {
      return {
        provide: TOKEN_MODULE_OPTIONS,
        useFactory: options.useFactory,
        inject: options.inject || []
      }
    }

    return {
      provide: TOKEN_MODULE_OPTIONS,
      useFactory: async (optionsFactory: TokenOptionsFactory) =>
        await optionsFactory.createTokenOptions(),
      inject: [options.useExisting || options.useClass!]
    }
  }
}
