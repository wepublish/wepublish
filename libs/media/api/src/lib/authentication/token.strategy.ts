import {Inject, Injectable, UnauthorizedException} from '@nestjs/common'
import {PassportStrategy} from '@nestjs/passport'
import {Strategy} from 'passport-http-bearer'

export const TOKEN_MODULE_OPTIONS = 'TOKEN_MODULE_OPTIONS'

export type TokenConfig = {
  token: string
}

@Injectable()
export class TokenStrategy extends PassportStrategy(Strategy, 'token') {
  constructor(@Inject(TOKEN_MODULE_OPTIONS) private config: TokenConfig) {
    super({
      session: false // would use a cookie if set to true
    })
  }

  public async validate(token: string) {
    if (token !== this.config.token) {
      throw new UnauthorizedException()
    }

    return true
  }
}
