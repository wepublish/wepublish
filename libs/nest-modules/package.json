{
  "name": "@wepublish/nest-modules",
  "version": "9.2.0",
  "description": "Nestjs API modules for we.publish.",
  "keywords": [
    "nestjs",
    "publishing",
    "media",
    "newspaper"
  ],
  "homepage": "https://github.com/wepublish/wepublish",
  "bugs": {
    "url": "https://github.com/wepublish/wepublish/issues",
    "email": "development@wepublish.ch"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/wepublish/wepublish.git",
    "directory": "libs/nest-modules"
  },
  "license": "MIT",
  "publishConfig": {
    "directory": "../../dist/libs/nest-modules"
  }
}
