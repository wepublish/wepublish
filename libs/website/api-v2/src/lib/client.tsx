import {
  ApolloClient,
  InMemoryCache,
  ApolloProvider,
  ApolloLink,
  NormalizedCacheObject
} from '@apollo/client'
import {ComponentType, memo, useMemo} from 'react'
import {BatchHttpLink} from '@apollo/client/link/batch-http'

export const getV2ApiClient = (
  apiUrl: string,
  links: ApolloLink[],
  cache?: NormalizedCacheObject
) => {
  const httpLink = new BatchHttpLink({
    uri: `${apiUrl}/v2`,
    batchMax: 5,
    batchInterval: 20
  })
  const link = [...links, httpLink].reduce(
    (links: ApolloLink | undefined, link) => (links ? links.concat(link) : link),
    undefined
  )

  return new ApolloClient({
    link,
    cache: new InMemoryCache().restore(cache ?? {}),
    ssrMode: typeof window === 'undefined'
  })
}

export const useV2ApiClient = (
  apiUrl: string,
  links: ApolloLink[] = [],
  cache?: NormalizedCacheObject
) => useMemo(() => getV2ApiClient(apiUrl, links, cache), [apiUrl, links, cache])

export const createWithV2ApiClient =
  (apiUrl: string, links: ApolloLink[] = [], cache?: NormalizedCacheObject) =>
  <
    // eslint-disable-next-line @typescript-eslint/ban-types
    P extends object
  >(
    ControlledComponent: ComponentType<P>
  ) =>
    memo<P>(props => {
      const client = useV2ApiClient(apiUrl, links, cache)

      return (
        <ApolloProvider client={client}>
          <ControlledComponent {...props} />
        </ApolloProvider>
      )
    })
