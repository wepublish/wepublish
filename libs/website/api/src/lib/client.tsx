import {
  ApolloClient,
  InMemoryCache,
  ApolloProvider,
  ApolloLink,
  NormalizedCacheObject
} from '@apollo/client'
import {BatchHttpLink} from '@apollo/client/link/batch-http'

import {ComponentType, memo, useMemo} from 'react'

export const getV1ApiClient = (
  apiUrl: string,
  links: ApolloLink[],
  cache?: NormalizedCacheObject
) => {
  const httpLink = new BatchHttpLink({uri: `${apiUrl}/v1`, batchMax: 5, batchInterval: 20})
  const link = [...links, httpLink].reduce(
    (links: ApolloLink | undefined, link) => (links ? links.concat(link) : link),
    undefined
  )

  return new ApolloClient({
    link,
    cache: new InMemoryCache().restore(cache ?? {}),
    ssrMode: typeof window === 'undefined'
  })
}

export const useV1ApiClient = (
  apiUrl: string,
  links: ApolloLink[] = [],
  cache?: NormalizedCacheObject
) => useMemo(() => getV1ApiClient(apiUrl, links, cache), [apiUrl, links, cache])

export const createWithV1ApiClient =
  (apiUrl: string, links: ApolloLink[] = [], cache?: NormalizedCacheObject) =>
  <
    // eslint-disable-next-line @typescript-eslint/ban-types
    P extends object
  >(
    ControlledComponent: ComponentType<P>
  ) =>
    memo<P>(props => {
      const client = useV1ApiClient(apiUrl, links, cache)

      return (
        <ApolloProvider client={client}>
          <ControlledComponent {...props} />
        </ApolloProvider>
      )
    })
