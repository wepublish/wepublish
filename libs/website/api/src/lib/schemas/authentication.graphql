query Me {
  me {
    id
  }
}

mutation CreateSession($email: String!, $password: String!) {
  createSession(email: $email, password: $password) {
    user {
      email
    }

    token
  }
}
