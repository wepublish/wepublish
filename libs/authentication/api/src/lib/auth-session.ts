import {User, UserRole} from '@prisma/client'

export enum AuthSessionType {
  User = 'user',
  Token = 'token'
}

export type TokenSession = {
  type: AuthSessionType.Token
  id: string
  name: string
  token: string
  roles: UserRole[]
}

export type UserSession = {
  type: AuthSessionType.User
  id: string
  user: User
  roles: UserRole[]
  createdAt: Date
  expiresAt: Date
  token: string
}

export type AuthSession = TokenSession | UserSession
