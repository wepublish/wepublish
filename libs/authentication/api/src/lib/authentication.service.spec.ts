import {PrismaModule} from '@wepublish/nest-modules'
import {PrismaClient} from '@prisma/client'
import {TestingModule, Test} from '@nestjs/testing'
import {AuthenticationService} from './authentication.service'
import {AuthSession, AuthSessionType} from './auth-session'

describe('AuthenticationService', () => {
  let service: AuthenticationService
  let prisma: PrismaClient

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports: [PrismaModule],
      providers: [AuthenticationService]
    }).compile()

    prisma = module.get<PrismaClient>(PrismaClient)
    service = module.get<AuthenticationService>(AuthenticationService)
  })

  it('should return a token session', async () => {
    const sessionSpy = jest
      .spyOn(prisma.session, 'findFirst')
      .mockReturnValue(Promise.resolve(null) as any)
    const tokenSpy = jest.spyOn(prisma.token, 'findFirst').mockReturnValue(
      Promise.resolve({
        id: '1234-1234',
        name: 'Foo Token',
        token: '1234',
        roleIDs: ['1234', '12345']
      }) as any
    )
    const userRoleSpy = jest
      .spyOn(prisma.userRole, 'findMany')
      .mockReturnValue(Promise.resolve([]) as any)
    const userSpy = jest.spyOn(prisma.user, 'findUnique')

    const result = await service.getSessionByToken('1234')
    expect(result).toMatchSnapshot()
    expect(sessionSpy.mock.calls[0]).toMatchSnapshot()
    expect(tokenSpy.mock.calls[0]).toMatchSnapshot()
    expect(userRoleSpy.mock.calls[0]).toMatchSnapshot()
    expect(userSpy).not.toHaveBeenCalled()
  })

  it('should return a user session', async () => {
    const sessionSpy = jest.spyOn(prisma.session, 'findFirst').mockReturnValue(
      Promise.resolve({
        userID: '12345'
      }) as any
    )
    const tokenSpy = jest
      .spyOn(prisma.token, 'findFirst')
      .mockReturnValue(Promise.resolve(null) as any)
    const userRoleSpy = jest
      .spyOn(prisma.userRole, 'findMany')
      .mockReturnValue(Promise.resolve([]) as any)
    const userSpy = jest.spyOn(prisma.user, 'findUnique').mockReturnValue(
      Promise.resolve({
        roleIDs: ['1111', '2222']
      }) as any
    )

    const result = await service.getSessionByToken('1234')
    expect(result).toMatchSnapshot()
    expect(sessionSpy.mock.calls[0]).toMatchSnapshot()
    expect(tokenSpy.mock.calls[0]).toMatchSnapshot()
    expect(userRoleSpy.mock.calls[0]).toMatchSnapshot()
    expect(userSpy.mock.calls[0][0]).toMatchSnapshot({
      select: expect.any(Object)
    })
  })

  it("should return null if user can't be found", async () => {
    const sessionSpy = jest.spyOn(prisma.session, 'findFirst').mockReturnValue(
      Promise.resolve({
        userID: '12345'
      }) as any
    )
    const tokenSpy = jest
      .spyOn(prisma.token, 'findFirst')
      .mockReturnValue(Promise.resolve(null) as any)
    const userRoleSpy = jest
      .spyOn(prisma.userRole, 'findMany')
      .mockReturnValue(Promise.resolve([]) as any)
    const userSpy = jest
      .spyOn(prisma.user, 'findUnique')
      .mockReturnValue(Promise.resolve(null) as any)

    const result = await service.getSessionByToken('1234')
    expect(result).toBeNull()
    expect(sessionSpy.mock.calls[0]).toMatchSnapshot()
    expect(tokenSpy.mock.calls[0]).toMatchSnapshot()
    expect(userRoleSpy.mock.calls[0]).toMatchSnapshot()
    expect(userSpy.mock.calls[0][0]).toMatchSnapshot({
      select: expect.any(Object)
    })
  })

  it('should return that the session is valid if expiresAt is in the future', () => {
    const today = new Date()
    const future = new Date(today)
    future.setDate(future.getDate() + 5000)

    const session = {
      type: AuthSessionType.User,
      expiresAt: future
    } as AuthSession

    const result = service.isSessionValid(session)
    expect(result).toBeTruthy()
  })

  it("should return that the session is valid if it's a token session", () => {
    const session = {
      type: AuthSessionType.Token
    } as AuthSession

    const result = service.isSessionValid(session)
    expect(result).toBeTruthy()
  })

  it('should return that the session is invalid if expiresAt is in the past', () => {
    const today = new Date()
    const past = new Date(today)
    past.setDate(past.getDate() - 5000)

    const session = {
      type: AuthSessionType.User,
      expiresAt: past
    } as AuthSession

    const result = service.isSessionValid(session)
    expect(result).toBeFalsy()
  })

  it("should return that the session is invalid if it's null", () => {
    const result = service.isSessionValid(null)
    expect(result).toBeFalsy()
  })
})
