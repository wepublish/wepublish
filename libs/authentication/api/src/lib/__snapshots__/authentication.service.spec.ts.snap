// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthenticationService should return a token session 1`] = `
{
  "id": "1234-1234",
  "name": "Foo Token",
  "roles": [],
  "token": "1234",
  "type": "token",
}
`;

exports[`AuthenticationService should return a token session 2`] = `
[
  {
    "where": {
      "token": "1234",
    },
  },
]
`;

exports[`AuthenticationService should return a token session 3`] = `
[
  {
    "where": {
      "token": "1234",
    },
  },
]
`;

exports[`AuthenticationService should return a token session 4`] = `
[
  {
    "where": {
      "id": {
        "in": [
          "1234",
          "12345",
        ],
      },
    },
  },
]
`;

exports[`AuthenticationService should return a user session 1`] = `
{
  "createdAt": undefined,
  "expiresAt": undefined,
  "id": undefined,
  "roles": [],
  "token": undefined,
  "type": "user",
  "user": {
    "roleIDs": [
      "1111",
      "2222",
    ],
  },
}
`;

exports[`AuthenticationService should return a user session 2`] = `
[
  {
    "where": {
      "token": "1234",
    },
  },
]
`;

exports[`AuthenticationService should return a user session 3`] = `
[
  {
    "where": {
      "token": "1234",
    },
  },
]
`;

exports[`AuthenticationService should return a user session 4`] = `
[
  {
    "where": {
      "id": {
        "in": [
          "1111",
          "2222",
        ],
      },
    },
  },
]
`;

exports[`AuthenticationService should return a user session 5`] = `
{
  "select": Any<Object>,
  "where": {
    "id": "12345",
  },
}
`;

exports[`AuthenticationService should return null if user can't be found 1`] = `
[
  {
    "where": {
      "token": "1234",
    },
  },
]
`;

exports[`AuthenticationService should return null if user can't be found 2`] = `
[
  {
    "where": {
      "token": "1234",
    },
  },
]
`;

exports[`AuthenticationService should return null if user can't be found 3`] = `undefined`;

exports[`AuthenticationService should return null if user can't be found 4`] = `
{
  "select": Any<Object>,
  "where": {
    "id": "12345",
  },
}
`;
